include/alp/omp/storage.hpp

```diff
@@ -0,0 +1,798 @@
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ *
+ * @file
+ *
+ * This file registers mechanisms for coordinate mapping between
+ * logical and physical iteration spaces.
+ *
+ */
+
+#ifndef _H_ALP_OMP_STORAGE
+#define _H_ALP_OMP_STORAGE
+
+#include <cmath>
+
+#include <alp/amf-based/storage.hpp>
+
+namespace alp {
+
+	namespace internal {
+
+		/** Specialization for matrices */
+		template< typename Structure >
+		struct determine_poly_factory< Structure, imf::Id, imf::Id, omp > {
+
+			typedef storage::polynomials::FullFactory<> factory_type;
+		};
+
+		/** Specialization for vectors */
+		template< typename Structure >
+		struct determine_poly_factory< Structure, imf::Id, imf::Zero, omp > {
+
+			typedef storage::polynomials::ArrayFactory factory_type;
+		};
+
+	} // namespace internal
+
+
+	/**
+	 * Implements mapping between global and local iteration spaces
+	 * for shared-memory parallel backend.
+	 * The logical coordinates are represented as pair (i, j) of row and
+	 * column positions within the matrix.
+	 * The local coordinates are represented as (tr, tc, rt, br, bc, il, jl),
+	 * where:
+	 *  - tr is thread row-coordinate
+	 *  - tc is thread column-coordinate
+	 *  - rt is replication factor for thread coordinates
+	 *  - br is block row-coordinate
+	 *  - bc is block column-coordinate
+	 *  - i  is element's row-coordinate within its block
+	 *  - j  is element's column-coordinate within its block
+	 *
+	 * This implementation assumes block-cyclic distribution of blocks
+	 * among threads.
+	 *
+	 */
+	class Distribution {
```

Comment:
Should this be `BlockCyclicDistribution` ? `Distribution` might become parent if needed. Should define it within a namespace (eg, `distributions`)?